nibabel/tests/test_image_types.py

```diff
@@ -0,0 +1,181 @@
+
+    n1_hdr['magic'] = 'n+1'
+    assert_equal(wat(n1_hdr), Nifti1Header)
+
+    n1_hdr['magic'] = 'ni1'
+    assert_equal(wat(n1_hdr), Nifti1Header)
+
+
+def test_sniff_and_guessed_image_type(img_klasses=all_image_classes):
+    """
+    Loop over all test cases:
+      * whether a sniff is provided or not
+      * randomizing the order of image classes
+      * over all known image types
+
+    For each, we expect:
+       * When the file matches the expected class, things should
+            either work, or fail if we're doing bad stuff.
+       * When the file is a mismatch, the functions should not throw.
+    """
+
+    def test_image_class(img_path, expected_img_klass):
+        """ Compare an image of one image class to all others.
+
+        The function should make sure that it loads the image with the expected
+        class, but failing when given a bad sniff (when the sniff is used)."""
+
+        def check_img(img_path, img_klass, sniff_mode, sniff, expect_success,
+                      msg):
+            """Embedded function to do the actual checks expected."""
+
+            if sniff_mode == 'no_sniff':
+                # Don't pass any sniff--not even "None"
+                is_img, new_sniff = img_klass.is_image(img_path)
+            else:
+                # Pass a sniff, but don't reuse across images.
+                is_img, new_sniff = img_klass.is_image(img_path, sniff)
+
+            if expect_success:
+                # Check that the sniff returned is appropriate.
+                new_msg = '%s returned sniff==None (%s)' % (img_klass.__name__,
+                                                            msg)
+                expected_sizeof_hdr = getattr(img_klass.header_class,
+                                              'sizeof_hdr', 0)
+                current_sizeof_hdr = 0 if new_sniff is None else len(new_sniff)
+                assert_true(current_sizeof_hdr >= expected_sizeof_hdr, new_msg)
+
+                # Check that the image type was recognized.
+                new_msg = '%s (%s) image is%s a %s image.' % (
+                    basename(img_path),
+                    msg,
+                    '' if is_img else ' not',
+                    img_klass.__name__)
+                assert_true(is_img, new_msg)
+
+            if sniff_mode == 'vanilla':
+                return new_sniff
+            else:
+                return sniff
+
+        sizeof_hdr = getattr(expected_img_klass.header_class, 'sizeof_hdr', 0)
+
+        for sniff_mode, sniff in dict(
+                vanilla=None,  # use the sniff of the previous item
+                no_sniff=None,  # Don't pass a sniff
+                none=None,  # pass None as the sniff, should query in fn
+                empty='',  # pass an empty sniff, should query in fn
+                irrelevant='a' * (sizeof_hdr - 1),  # A too-small sniff, query
+                bad_sniff='a' * sizeof_hdr).items():  # Bad sniff, should fail.
```

Comment:
Shouldn't these be `bytes` objects (b'a')? They'll either fail or trigger a new sniff anyway, so maybe not a big deal, but if you want them to fail for the right reasons on Python 3...